spacy_llm/tasks/span.py

```diff
@@ -143,39 +196,76 @@ def assign_spans(
         """Assign spans to the document."""
         raise NotImplementedError()
 
+    def _extract_span_reasons(self, response: str) -> List[SpanReason]:
+        """Parse raw string response into a list of SpanReasons"""
+        span_reasons = []
+        for line in response.strip().split("\n"):
+            try:
+                span_reason = SpanReason.from_str(line)
+            except ValueError:
+                continue
+            if not span_reason.is_entity:
+                continue
+            norm_label = self._normalizer(span_reason.label)
+            if norm_label not in self._label_dict:
+                continue
+            label = self._label_dict[norm_label]
+            span_reason.label = label
+            span_reasons.append(span_reason)
+        return span_reasons
+
     def parse_responses(
         self, docs: Iterable[Doc], responses: Iterable[str]
     ) -> Iterable[Doc]:
-        for doc, prompt_response in zip(docs, responses):
+        """Since we provide entities in a numbered list, we expect the LLM to
```

Comment:
The way we parse the responses needs to change entirely. 

We now assume the entities are reported from the LLM in order and we map entities/spans incrementally based utilizing the `find_after` param of `find_substrings`. 

So for input text: 

```
"The Denver Nuggets beat the Miami Heat 4 games to 1"
```

and a response:

```
1. Denver Nuggets | True | BASKETBALL_TEAM | is a basketball team in the NBA
2. Miami Heat | True | BASKETBALL_TEAM | is a basketball team in the NBA
3. 4 | False | ==NONE== | is a number
4. games | False | ==NONE== | is a common noun
5. 1 | False | ==NONE== | is a number
```


we first parse the list of `SpanReason`s which is easy, break by newline and call `SpanReason.from_str(line)`.

When we map back to spaCy spans now, we have some ordered structured so we no longer need to do `single_match=False` (and in fact that setting leads to false positives so it's good we're removing it)

So for ent 1, we match against the whole input text.

1. Matching `Denver Nuggets`

Text to match against
```
"The [Denver Nuggets](Found) beat the Miami Heat 4 games to 1"
```

Now we consider only the text after this:

For NER, we find after the `end_char` of the last found span, for SpanCat, we consider overlapping entities so we use the `start_char`.

2. Matching "Miami Heat"

Text to match against:
```
" beat the [Miami Heat](FOUND) 4 games to 1"
```


This incremental parsing of the span reasons against a smaller and smaller section of the input string works great most of the time. Issues arise with this specific test case and Spancat.
https://github.com/explosion/spacy-llm/pull/180/files#r1279693295

Reply:
The big open question here now is should we prioritize `case_sensitive` matches? I think we should. I think if the case is correct in the LLM response and we can find a match let's use that. Then fallback to `case_insensitive` only if no case sensitive matches are found for a line of output.

So for input text:

```
This guy jean (or Jean) is the president of the Jean Foundation.
```

And output:

```
1. Jean | True | PER | is a person's name
```

with `case_sensitive=False`, we would have marked:

```
This guy [jean](PERSON) (or Jean) is the president of the Jean Foundation.
```

But I think first we should scan for case sensitive matches, then fallback afterward.

So we'd get a final output of:

```
This guy jean (or [Jean](PERSON)) is the president of the Jean Foundation.
```

regardless of whether `case_sensitive` is True or False.

Reply:
@svlandeg @rmitsch if you could weigh in here that would be awesome!

Reply:
- Incremental parsing seems reasonable.
- Same for prioritizing case-sensitive matching.
- Commenting on the test case further above.

Reply:
Sweet, I'll keep these changes as is then :)

Reply:
Hm, I'm not sure I agree. If `case_sensitive` is set specifically to `False`, then there shouldn't be some hidden logic that still prioritizes case-matching variants.

Reply:
I do think that in general the way results are parsed now seems much better - i.e. requiring an order and needing two SpanReasons rather than relying on a "match all" or "match once" logic. 

On our internal board, we had a similar example that we wanted to fix, but that we had blocked on this PR because I assumed this PR would solve it:

In the sentence "The woman Paris was walking around in Paris, talking to her friend Paris" we'd need the LLM to return something like

Paris: PER
Paris: LOC
Paris: PER

in order, to be able to parse this example correctly.

Reply:
Hmm ok I see your point, I could remove the separate hidden case sensitive match step. That's probably for the best. As for order of extraction, I saw that issue in GitHub already and added a specific regression test for this sentence: https://github.com/explosion/spacy-llm/pull/180/files#r1282197256

Reply:
@kabirkhan : the regression test has only two instances though, I wanted to test specifically with two PERSON entities and a LOC inbetween. Can the current implementation deal with that?

Reply:
Ok well actually, the issue is this sentence referenced above doesn't work without the case sensitive matching step and the way I'm using the `find_after` param with spancat.

Input:
```
This guy jean (or Jean) is the president of the Jean Foundation.
```

Response:

```
1. jean | True | PER | is a person's name
2. Jean | True | PER | is a person's name
3. Jean Foundation | True | ORG | is the name of an Organization name
```


### Approach without just case insensitive matching

* Will find the first "jean" and mark the `find_after` parameter to the `start_char` of "jean" because we want to allow overlapping spans.
* Will then look for "Jean" but case insensitive ("jean") and since the parsing starts at the `start_char` (not the `end_char` of the first "jean", we mark that span again, then move on to the next entity "Jean Foundation"
* "Jean Foundation" gets parsed correctly

Final spans list is: `["jean", "jean", "Jean Foundation"]

We end up doing a `set(spans)` to remove identical spans so we only get 2 spans returned from this response of 3 spans.

This is only for Spancat with `allow_overlap=True`. NER works correctly since we set the `find_after` to the end of the previously found ent. If we did this for `Spancat` we'd never parse out overlapping spans.

Reply:
> @kabirkhan : the regression test has only two instances though, I wanted to test specifically with two PERSON entities and a LOC inbetween. Can the current implementation deal with that?

yep! as long as the LLM responds with all 3 entities the parsing will work for NER or for SpanCat with allow_overlap=False. If allow_overlap is True for Spancat, the current implementation won't work. 

Also I fixed the issue with the "jean", "Jean", "Jean Foundation" parsing in the latest commit.

Reply:
The only way to get that sentence to work with `SpanCat` and `allow_overlap=True` would be to alter the prompt in some way that LLM marks a position in the text when it extracts SpanReason's. Having the LLM respond with `start_char` and `end_char` didn't really work at all when I tried. 

I definitely think it's something we could try to address in another new version of the model with something like this Paper's implementation: https://arxiv.org/abs/2304.10428 where they mark offsets in the original text

Reply:
> I definitely think it's something we could try to address in another new version of the model with something like this Paper's implementation: https://arxiv.org/abs/2304.10428 where they mark offsets in the original text

I use this kind of in-text markers for the EL task, and it seems to work quite well (still have to run experiments to get numbers).

> The only way to get that sentence to work with SpanCat and allow_overlap=True would be to alter the prompt in some way that LLM marks a position in the text when it extracts SpanReason's. Having the LLM respond with start_char and end_char didn't really work at all when I tried.

I would have assumed that with `SpanCat` `allow_overlap is True` always applies? Anyway, if that's a limitation of the current approach, that's ok. We can fix that/use in-text markers in a later PR, but let's wrap this up.

Reply:
I think we can close this?

Reply:
Do we now have a unit test for the original example I mentioned (with the three entities, first PER, then LOC, then PER again)?

Reply:
No, I'll add one.

Reply:
See https://github.com/explosion/spacy-llm/pull/180/commits/2979ffc9f5e077ffc5f84619ad918ddabe7c95d0. Had to make some adjustments to the config to make the example work.

Reply:
Awesome, thanks!